src/Lean/Meta/Match/MatcherApp/Transform.lean

```diff
@@ -156,4 +156,219 @@ def refineThrough? (matcherApp : MatcherApp) (e : Expr) :
   catch _ =>
     return none
 
+
+/--
+Given `n` and a type `α₁ → α₂ → ... → αₙ → Sort u`, returns the types `α₁, α₂, ..., αₙ`.
+
+This can be used to infer the expected type of the alternatives when constructing a `MatcherApp`.
+-/
+-- TODO: Which is the natural module for this?
+def arrowDomainsN (n : Nat) (type : Expr) : MetaM (Array Expr) := do
+  let mut type := type
+  let mut ts := #[]
+  for i in [:n] do
+    type ← whnfForall type
+    let Expr.forallE _ α β _ ← pure type | throwError "expected {n} arguments, got {i}"
+    if β.hasLooseBVars then throwError "unexpected dependent type"
+    ts := ts.push α
+    type := β
+  return ts
+
+/--
+Performs a possibly type-changing transformation to a `MatcherApp`.
+
+* `onParams` is run on each parameter and discrimitant
+* `onMotive` runs on the body of the motive, and is passed the motive parameters
+  (same number as discrs)
```

Comment:
What's a "discr"? Same as last time, or not? It's good to be consistent with abbreviations.

Reply:
```suggestion
  (one for each `MatcherApp.discrs`)
```